grin/src/AbstractInterpretation/CByUtil.hs

```diff
@@ -0,0 +1,181 @@
+
+fromProducerGraph :: ProducerGraph -> ProducerGraph'
+fromProducerGraph = Map.map _producerSet . _producerMap . _producerGraph
+
+collectActiveProducers :: LVAResult -> ProducerMap -> Set Name
+collectActiveProducers lvaResult = selectActiveProducers lvaResult . collectProducers
+
+collectProducers :: ProducerMap -> Set Name
+collectProducers = mconcat
+                 . concatMap Map.elems
+                 . Map.elems
+                 . Map.map _producerSet
+                 . _producerMap
+
+
+-- Selects the active producers from a producer set.
+-- A producers is active if at least one of its tags has a live field.
+selectActiveProducers :: LVAResult -> Set Name -> Set Name
+selectActiveProducers lvaResult prods = Map.keysSet
+                                      . Map.filter hasActiveTag
+                                      . Map.map nodeLiveness
+                                      . producerLiveness
+                                      $ lvaResult
+  where
+
+  producerLiveness :: LVAResult -> Map Name Liveness
+  producerLiveness = flip Map.restrictKeys prods . _register
+
+  nodeLiveness :: Liveness -> Map Tag (Vector Bool)
+  nodeLiveness (NodeSet m) = Map.map _node m
+  nodeLiveness _ = error "Producers cannot have non-node liveness information"
+
+  hasActiveTag :: Map Tag (Vector Bool) -> Bool
+  hasActiveTag = any (Vec.elem True) . Map.elems
+
+-- Constructs the basic connection graph between all producers.
+-- If a consumer has multiple producers with the same tag,
+-- then one producer will be selected, and the others will be connected to it.
+mkBasicProdGraph :: ProducerMap -> ProducerGraph'
+mkBasicProdGraph producers = flip execState mempty $ do
+  let
+    -- All the active producers found in the program grouped by tags.
+    taggedGroups :: [(Tag, Set Name)]
+    taggedGroups = concatMap (Map.toList . _producerSet)
+                 . Map.elems
+                 . _producerMap
+                 $ producers
+
+  forM taggedGroups $ \(t,ps) -> do
+    let (p:_)  = Set.toList ps
+        entry  = Map.singleton t ps
+        update = Map.unionWith Set.union
+    modify $ Map.insertWith update p entry
+
+-- Deletes all connections then connects each producer with itself
+onlyReflexiveConnections :: ProducerGraph' -> ProducerGraph'
+onlyReflexiveConnections = Map.mapWithKey (\k m -> Map.map (const $ Set.singleton k) m)
+
+-- Creates an undirected graph from a directed one by connecting vertices
+-- in both directions. Also connects each vertex with itself.
+undirectedReflexiveClosure :: ProducerGraph' -> ProducerGraph'
+undirectedReflexiveClosure m = flip execState m $ do
+  let pList = Map.toList
+            . Map.map Map.toList
+            . Map.map (Map.map Set.toList)
+            $ m
+  -- for each (p, (t, [p1 .. pn])),
+  -- it add the entries: (p1, (t, [p])) .. (pn, (t, [p]))
+  -- also insert p into (p, (t, [p1 .. pn])),
+  forM pList $ \(p, taggedGroups) ->
```

Comment:
This feels like very imperative style. You could generate the lazy list and single mapM on that.